### ISSUE TITLE
Move crossgen out from the tools dir and into the lib dir instead

### ISSUE THREAD

**Issue by joperezr** [dotnet/coreclr#3649]
Currently, the CoreCLR package puts the crossgen program under the tools directory, and because of this dotnet publish won't copy it over to the runtime. The reason why we need it is because in corefx we want to ngen the runtime we use to build the repo. Is it posible to change the nuspec to have the crossgen program under lib?

related: dotnet/buildtools#487

cc: @JohnChen0 @weshaggard

**Comment by gkhanna79** [comment]
I concur with @weshaggard. Please see how this is done in CLI since they endup carrying crossgen as part of the CLI distribution. @brthor maybe able to help you.

**Comment by joperezr** [comment]
I guess we could do it how CLI does it but I don't really like [their approach](https://github.com/dotnet/cli/blob/88c1340f6d49bb8275df5a6f69f62ace54bd590e/scripts/dotnet-cli-build/CompileTargets.cs#L331) since basically what they do is to hardcode the location of the package that was restored and manually copy it from there into the runtime. I feel that this would work in most cases, but there would always be a scenario in which if some env vars are not set in one machine this would fail. If you guys still feel that this is not convincing enough, then I'm fine implementing it like cli does.

**Comment by weshaggard** [comment]
@davidfowl @anurse do you guys know of a good way to publish the "tools" directory from a package?We are trying to get the tools from some packages as we setup our toolset and so we essentially restore and publish on the set of packages but today we don't have a good way to get the assets out of the tools directory in the packages without hardcoding the package versions in the paths.

**Comment by brthor** [comment]
Unfortunately this hardcoding is necessary because crossgen lives in the "tools" directory of the package. There's no nuget support for that directory, and at that point the .nupkg needs to be treated like a zip file, extracting what we need with paths.

ilc.exe (CoreRT) gets around this by being placed in the "native" directory. Like you mentioned @weshaggard this does cause ilc.exe to get published to the output directory whenever you use `dotnet publish`. If that behavior is undesired for anyone publishing the runtime package, perhaps crossgen.exe could be in a separate package.

**Comment by rahku** [comment]
this also affects coreclr helix runs. There are several tests which require crossgen to be copied to the runtime folder. But it does not get copied as crossgen is in tools dir. I was talking to @ericstj about this. He suggested that crossgen should be put in a separate package and should be removed from tools dir of runtime package.

**Comment by weshaggard** [comment]
Yes another crossgen package which contains the assets in a native folder would do the trick. @gkhanna79 is this something you think would work?

**Comment by gkhanna79** [comment]
Before we go this route, I would like to understand if creating a new package, for anything not under native folder in nuget package, is the standard/formal guidance for restoring it for consumption. IMHO, this looks to be an overkill - is nuget intending to support such scenarios in an upcoming release as I am sure this is a common thing to do and we cannot be creating packages for all such artifacts?

**Comment by gkhanna79** [comment]
I dont believe this is required for RTM, especially since there overall picture for tools in nuget packages is not understood. Please let me know if there are concerns moving this out of RTM.

**Comment by joperezr** [comment]
agreed, I'll move it to Future for now.

**Comment by joperezr** [comment]
Are we still planning on doing this eventually or not anymore? if not, then I can close this issue.

**Comment by gkhanna79** [comment]
It is marked for us to look into for future.
